src/main/java/io/cdap/plugin/spark/HierarchyFlattener.java

```diff
@@ -347,159 +350,45 @@ public HierarchyFlattener(HierarchyConfig config) {
-      } else {
-        columns[i++] = input.col(fieldName);
-      }
-    }
-
-    return levelZero.union(input.select(columns)).distinct();
-  }
-
-  /*
-    Get roots by looking for rows where they parent never rows up as a child in any other row.
-   */
-  private Dataset<Row> getNonSelfReferencingRoots(Dataset<Row> input, List<String> dataFieldNames) {
+  private Dataset<Row> getStartingPoints(Dataset<Row> input, List<String> dataFieldNames,
+                                         @Nullable String condition) {
     /*
-       These types of roots are rows where the parent never shows up as a child in any other row.
-       They are found by running the following query:
+       Roots are found by running the following query:
 
        select
          A.parent as parent,
          A.parent as child,
          0 as level,
-         true as root,
          false as leaf,
          A.datafield1 as datafield1,
          ...,
          null as datafieldN
-       from input as A left outer join (select child from input) as B on A.parent = B.child
-       where B.child is null
+       from input
+       if condition != null : where <condition>
 
-       Whether a datafield is null or not depends on whether there is a mapping for that field.
-       For example, with mapping parentCategory -> category and input rows:
-
-         [parent:1, child:2, parentCategory:plants, category:vegetable, sold:100]
-         [parent:2, child:3, parentCategory:vegetable, category:onion, sold:50]
-
-       This results in:
-
-         [parent:1, child:1, level:0, root:true, leaf:false, parentCategory:plants, category:plants, sold:null]
-
-       parentCategory and category both get their value from the parentCategory key in the mapping.
-       Every other data field is just null.
     */
-    Column[] columns = new Column[dataFieldNames.size() + 5];
-    columns[0] = new Column("A." + parentCol).as(parentCol);
-    columns[1] = new Column("A." + parentCol).as(childCol);
-    columns[2] = functions.lit(0).as(levelCol);
-    columns[3] = functions.lit(trueStr).as(topCol);
-    columns[4] = functions.lit(0).as(botCol);
-    int i = 5;
-    Map<String, String> childParentMapping = new HashMap<>();
-    for (Map.Entry<String, String> entry : parentChildMapping.entrySet()) {
-      childParentMapping.put(entry.getValue(), entry.getKey());
+    Column[] columns = getLevel1Columns(dataFieldNames);
+    if (condition != null) {
+      return input.alias("A").where(condition).select(columns);
     }
-    for (String fieldName : dataFieldNames) {
-      /*
-         If this is a mapped field, the child field should be used as both the parent and child.
-         For example, suppose a root row looks like:
-
-           [parent:1, child:2, parentCategory:plants, category:vegetable, sold:100]
-
-         and there is a mapping of parentCategory -> category
-         that means the result should be:
+    return input.alias("A").select(columns);
```

Comment:
should we dedupe?

Reply:
I don't think we need to dedupe here because we dedupe at the end anyway to remove duplicate records emitted by intermediate join steps. https://github.com/data-integrations/hierarchical-to-relational/pull/8/files#diff-ce85d294002d4d5c22c1d8394bfb908c69aad1872c9d665d7401185b270cb2f9R323-R342